src/main/java/fopbot/Direction.java

```diff
@@ -1,23 +1,155 @@
+
+    /**
+     * Gets the y offset of the {@link Direction}.
+     *
+     * @return The y offset of the {@link Direction}.
+     */
+    public int getDy() {
+        return dy;
+    }
+
+    /**
+     * Checks whether this direction is a horizontal direction.
+     *
+     * @return {@code true} if this direction is a horizontal direction.
+     */
+    public boolean isHorizontal() {
+        return dy == 0;
+    }
+
+    /**
+     * Checks whether this direction is a vertical direction.
+     *
+     * @return {@code true} if this direction is a vertical direction.
+     */
+    public boolean isVertical() {
+        return dx == 0;
+    }
+
+    /**
+     * Calculates the length of a given vector of an arbitrary dimension.
+     *
+     * @param components the components of the vector.
+     * @return the length of the vector.
+     */
+    private double vectorLength(final double... components) {
+        return Math.sqrt(Arrays.stream(components).map(c -> Math.pow(c, 2)).sum());
+    }
+
+    /**
+     * Returns the direction who best matches the given vector.
+     *
+     * @param x The x-component of the vector.
+     * @param y The y-component of the vector.
+     * @return Returns the direction who best matches the given vector.
+     */
+    public Direction fromVector(final double x, final double y) {
+        if (x == 0 && y == 0) {
+            return NONE;
+        }
+        return Arrays
+            .stream(values())
+            .filter(d -> d != NONE)
+            .min(
+                // calculate the angle between the given vector and the direction vector
+                (d1, d2) -> Double.compare(
+                    Math.acos(
+                        (d1.dx * x + d1.dy * y) / (vectorLength(d1.dx, d1.dy) * vectorLength(x, y))
+                    ),
+                    Math.acos(
+                        (d2.dx * x + d2.dy * y) / (vectorLength(d2.dx, d2.dy) * vectorLength(x, y))
+                    )
+                )
+            )
+            .orElse(NONE);
+    }
+
+    /**
+     * Returns the direction opposite to this direction.
+     *
+     * @return the direction opposite to this direction.
+     */
+    public Direction getOpposite() {
+        return fromVector(-dx, -dy);
+    }
+
+    /**
+     * Checks whether the given direction is opposite to this direction.
+     *
+     * @param direction The direction to check.
+     * @return {@code true} if the given direction is opposite to this direction.
+     */
+    public boolean isOpposite(final Direction direction) {
+        return equals(direction.getOpposite());
+    }
+
+    /**
+     * Returns the direction that is the result of rotating this direction by the given angle in degrees counterclockwise.
+     *
+     * @param angle The angle to rotate this direction by.
+     * @return The direction that is the result of rotating this direction by the given angle in degrees counterclockwise.
+     */
+    public Direction rotate(final double angle) {
```

Comment:
I don't really see the point of this method, modulo math should be the way to work with directions

Reply:
will be removed